# Анализ сервиса для чтения книг


# Цель:

Основная цель проекта это создание ценностное предложение для нового продукта.


# Задача:

- Посчитать сколько книг вышло после 1 января 2000 года;

- Для каждой книги посчитать количество обзоров и среднюю оценку;

- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так можно исключить из анализа брошюры;

- Определить автора с самой высокой средней оценкой книг — учитывая только книги с 50 и более оценками;

- Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.


# Аббревиатуры:

## Датасет *books*:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

## Датасет *authors*:

- `author_id` — идентификатор автора;
- `author` — имя автора.

## Датасет *publishers*:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

## Датасет *ratings*:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

## Датасет *reviews*:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

# Оглавление:

[Вниз](#eight)<a id= 'nine'></a>

[Шаг 1. Изучение таблиц](#one)

[Шаг 2. Сколько книг вышло после 1 января 2000 года](#two)

[Шаг 3. Количество обзоров и средняя оценка для каждой книги](#three)

[Шаг 4. Определить издательство, которое выпустило наибольшее число книг](#four)

[Шаг 5. Определить автора с самой высокой средней оценкой книг](#five)

[Шаг 6. Посчитать среднее количество обзоров от пользователей](#six)

[Общий вывод](#seven)

<a id= 'one'></a>
# Шаг 1. 
# Изучение таблиц

In [1]:
import pandas as pd
import requests
from bs4 import BeautifulSoup
from sqlalchemy import create_engine 

In [2]:
db_config = {'user': 'praktikum_student', # имя пользователя
 'pwd': 'Sdf4$2;d-d30pp', # пароль
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
 'port': 6432, # порт подключения
 'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [3]:
books = ''' SELECT * FROM books'''

In [4]:
books = pd.io.sql.read_sql(books, con = engine)

In [5]:
books.to_csv('data.csv', index = False)

In [6]:
books

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [7]:
authors = ''' SELECT * FROM authors'''
authors = pd.io.sql.read_sql(authors, con = engine)
authors.to_csv('data.csv', index = False)

In [8]:
authors

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

In [9]:
ratings = ''' SELECT * FROM ratings'''
ratings = pd.io.sql.read_sql(ratings, con = engine)
ratings.to_csv('ratings.csv', index = False)

In [10]:
ratings

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

In [11]:
reviews = ''' SELECT * FROM reviews'''
reviews = pd.io.sql.read_sql(reviews, con = engine)
reviews.to_csv('reviews.csv', index = False)

In [12]:
reviews

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

<a id= 'two'></a>
# Шаг 2. 
# Сколько книг вышло после 1 января 2000 года

In [13]:
books_date = ''' 
    SELECT DISTINCT COUNT(*) 
    FROM books
    WHERE publication_date > '2000-01-01'
    '''

In [14]:
pd.io.sql.read_sql(books_date, con = engine)

count
0    819

После 2000 года вышло 819 книг, учитывая в общое количества книг равна 1000, то это довольно много.

<a id= 'three'></a>
# Шаг 3. 
# Количество обзоров и средняя оценка для каждой книги

In [15]:
reviews

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

In [16]:
book_review ='''
    SELECT 
        books.book_id as book_id,
        books.title as title,
        ROUND(AVG(ratings.rating)) as avg_rating,
        COUNT(DISTINCT reviews.text) as qnt_reviews
    FROM
        books
        LEFT JOIN ratings ON ratings.book_id = books.book_id
        LEFT JOIN reviews ON reviews.book_id = books.book_id
    GROUP BY
        books.book_id,
        books.title
    ORDER BY
        qnt_reviews DESC'''

In [17]:
pd.io.sql.read_sql(book_review, con = engine)

book_id                                              title  avg_rating  \
0        948                            Twilight (Twilight  #1)         4.0   
1        963                                Water for Elephants         4.0   
2        734                                   The Glass Castle         4.0   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...         4.0   
4        695  The Curious Incident of the Dog in the Night-Time         4.0   
..       ...                                                ...         ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel         4.0   
996      808                            The Natural Way to Draw         3.0   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites         5.0   
998      221                          Essential Tales and Poems         4.0   
999      191  Disney's Beauty and the Beast (A Little Golden...         4.0   

     qnt_reviews  
0              7  
1              6  
2              6  
3              6  
4              6  
..           ...  
995            0  
996            0  
997            0  
998            0  
999            0  

[1000 rows x 4 columns]

Я вывел книги у которых самое большое количество отзывов. Чем больше отзывов тем объективней средняя оценка, есть книги где средняя оценка равна 5 но отзывов нет, а есть книги как "Сумерки" где 7 отзывов и средняя оценка 4.

<a id= 'four'></a>
# Шаг 4. 
# Определить издательство, которое выпустило наибольшее число книг толще 50 страниц 

In [18]:
book_publisher ='''
    SELECT 
        COUNT(books.book_id) as num_books,
        publishers.publisher as publisher
    FROM
        books
        INNER JOIN publishers ON publishers.publisher_id = books.publisher_id
    WHERE
        books.num_pages > 50
    GROUP BY
        publishers.publisher
    ORDER BY
        num_books DESC'''

In [19]:
pd.io.sql.read_sql(book_publisher, con = engine)

num_books                                          publisher
0           42                                      Penguin Books
1           31                                            Vintage
2           25                           Grand Central Publishing
3           24                                   Penguin Classics
4           19                                   Ballantine Books
..         ...                                                ...
329          1                                         Turtleback
330          1  Atheneum Books for Young Readers: Richard Jack...
331          1                                     Penguin Signet
332          1                                    Victor Gollancz
333          1                      Harvard Business Review Press

[334 rows x 2 columns]

Самым успешным издетельством  в плане издательства книг это "Penguin Books". Таким образом, лучше всего сотрудничать с этим издательством.

<a id= 'five'></a>
# Шаг 5. 
# Определить автора с самой высокой средней оценкой книг

In [20]:
book_author ='''
    SELECT 
        authors.author as author,
        AVG(DF.avg_rating) as avg_rating,
        COUNT(DF.qnt_rating) as qnt_rating
    FROM
        books
        INNER JOIN(
            SELECT
                book_id as book_id,
                COUNT( DISTINCT ratings.rating_id) as qnt_rating,
                AVG(rating) as avg_rating
            FROM
                ratings
            GROUP BY
                book_id
            HAVING
                COUNT(ratings.rating_id) >= 50) AS DF ON DF.book_id = books.book_id

        INNER JOIN authors ON authors.author_id = books.author_id
    GROUP BY
        author
    ORDER BY
        avg_rating DESC
    LIMIT
        25
    '''

In [21]:
pd.io.sql.read_sql(book_author, con = engine)

author  avg_rating  qnt_rating
0                          J.K. Rowling/Mary GrandPré    4.283844           4
1                   Markus Zusak/Cao Xuân Việt Khương    4.264151           1
2                                      J.R.R. Tolkien    4.258446           2
3                                   Louisa May Alcott    4.192308           1
4                                        Rick Riordan    4.080645           1
5                                     William Golding    3.901408           1
6                                       J.D. Salinger    3.825581           1
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce    3.789474           1
8   William Shakespeare/Paul Werstine/Barbara A. M...    3.787879           1
9                                           Dan Brown    3.754540           2
10                                         Lois Lowry    3.750000           1
11            George Orwell/Boris Grabnar/Peter Škerl    3.729730           1
12                                    Stephenie Meyer    3.662500           1
13                                     John Steinbeck    3.622951           1

Вывел всех писателей у которых книги набрали более 50ти оценок и у всех оказалась средняя оценка около 4. Я решил распределить писателей по средней оценке, где на первом месте оказались J.K. Rowling/Mary GrandPré. Чем больше оценок тем объективней будет средняя оценка.

<a id= 'six'></a>
# Шаг 6. 
# Посчитать среднее количество обзоров от пользователей  которые поставили больше 50 оценок

In [22]:
user_review ='''
    SELECT 
        ROUND(AVG(subquery.qnt_reviews)) as avg_reviews
        
    FROM
        ratings
        LEFT JOIN(
            SELECT
                COUNT(review_id) as qnt_reviews,
                username as username
            FROM
                reviews
            GROUP BY
                username
                ) AS subquery ON subquery.username = ratings.username
    HAVING
        COUNT(ratings.rating_id) > 50
    LIMIT
        10
    '''

In [23]:
pd.io.sql.read_sql(user_review, con = engine)

avg_reviews
0         18.0

Среднее количество обзоров от пользователей которые поставили больше 50 оценок равна 18.

<a id= 'seven'></a>
# Общий Вывод

После анализа данных с книгами можно выделить следующие моменты:

- Я смог подсчитать количество выведших книг после 2000 года
- Теперь можно увидеть какие книги стоит в первую очередь предлагать новым клиентом, так как они являются самыми популярными в нашей базе
- В дальнейшем стоит сотрудничать с такими издательствами как Penguin Books и Vintage
- Теперь есть список самых высоко оцененных писателей.

Эти все данные должны помочь в формировании предложения для новых пользователей, и выбрать для них наиболее популярные и высоко оцененные книги.

[Наверх](#nine)<a id= 'eight'></a>